pkg/services/meta/meta.go

```diff
@@ -0,0 +1,125 @@
+package meta
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/nspcc-dev/neo-go/pkg/core/block"
+	"github.com/nspcc-dev/neo-go/pkg/core/state"
+	"github.com/nspcc-dev/neo-go/pkg/rpcclient"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	"go.uber.org/zap"
+)
+
+const (
+	// raw storage prefixes.
+
+	// rootKey is the key for the last known state root in KV data base
+	// associated with MPT.
+	rootKey = 0x00
+)
+
+// ContainerLister is a source of actual containers current node belongs to.
+type ContainerLister interface {
+	// List returns node's containers and
+	// any error that does not allow listing.
+	List() (map[cid.ID]struct{}, error)
+}
+
+// Meta handles object meta information received from FS chain and object
+// storages. Chain information is stored in Merkle-Patricia Tries. Full objects
+// index is built and stored as a simple KV storage.
+type Meta struct {
+	l        *zap.Logger
+	stopCh   chan struct{}
+	rootPath string
+
+	netmapH util.Uint160
+	cnrH    util.Uint160
+	cLister ContainerLister
+
+	ws      *rpcclient.WSClient
+	bCh     chan *block.Header
+	objEv   chan *state.ContainedNotificationEvent
+	cnrEv   chan *state.ContainedNotificationEvent
+	epochEv chan *state.ContainedNotificationEvent
+
+	m        sync.RWMutex
+	storages map[cid.ID]*containerStorage
+}
+
+// New makes [Meta].
+func New(l *zap.Logger, cLister ContainerLister, ws *rpcclient.WSClient, containerH, nmH util.Uint160, rootPath string) (*Meta, error) {
+	storagesFS, err := os.ReadDir(rootPath)
```

Comment:
This limits us to less than 1024 containers by default, btw.

Reply:
can you elaborate? what exactly limits us? we can add a tree structure if needed. but i am not sure i understood our point

Reply:
`ulimit -n`. It can be changed, not something set in stone, but still.